pkg/controller/drain.go

```diff
@@ -373,6 +387,236 @@ func (o *DrainOptions) evictPods(pods []api.Pod, policyGroupVersion string, getP
+	volMap := make(map[string][]string)
+	for _, pod := range pods {
+		podPVs := o.getPvs(pod)
+		volMap[pod.Namespace+"/"+pod.Name] = podPVs
+	}
+	glog.V(4).Info("Volume map: ", volMap)
+	return volMap
+}
+
+func filterSharedPVs(volMap map[string][]string) {
+	sharedVol := make(map[string]bool)
+	sharedVolumesFound := false
+
+	// Create hash map of volumes:
+	// Key: volume name
+	// Value: 'true' if any other pod shares this volume, else 'false'
+	for _, vols := range volMap {
+		for _, vol := range vols {
+			if _, ok := sharedVol[vol]; !ok {
+				sharedVol[vol] = false
+			} else {
+				sharedVol[vol] = true
+				sharedVolumesFound = true
+			}
+		}
+	}
+
+	if !sharedVolumesFound {
+		glog.V(4).Info("No shared volumes found.")
+		return
+	}
+
+	// Recreate the values of volMap. Append volume if it is not shared
+	for pod, vols := range volMap {
+		volList := []string{}
+		for _, vol := range vols {
+			if sharedVol[vol] == false {
+				volList = append(volList, vol)
+			}
+		}
+		volMap[pod] = volList
+	}
+	glog.V(4).Info("Removed shared volumes. Filtered list of pods with volumes: ", volMap)
+}
+
+func (o *DrainOptions) evictPodsWithPv(pods []corev1.Pod, policyGroupVersion string, getPodFn func(namespace, name string) (*api.Pod, error), returnCh chan error) {
+	returnChInt := make(chan error, 1)
+	detachCh := make(chan bool, 1)
+
+	sortPodsByPriority(pods)
+
+	volMap := o.doAccountingOfPvs(pods)
+	filterSharedPVs(volMap)
+
+	for {
+		retryPods := []corev1.Pod{}
+		for _, pod := range pods {
+			pvs := volMap[pod.Namespace+"/"+pod.Name]
+			go o.evictPodInternal(pod, policyGroupVersion, getPodFn, returnChInt, true)
+			startTime := time.Now()
+			select {
+			case err := <-returnChInt:
+				if apierrors.IsTooManyRequests(err) {
+					glog.V(3).Info("Pod ", pod.Namespace, "/", pod.Name, " couldn't be evicted because of PDB violation. Will be retried.")
+					// Pod eviction failed because of PDB violation, retry
+					retryPods = append(retryPods, pod)
+					continue
+				} else if err != nil {
+					returnCh <- fmt.Errorf("Error when evicting pod: %v/%v. Err: %v", pod.Namespace, pod.Name, err)
+					continue
+				}
+
+				go o.waitForDetach(pvs, o.nodeName, detachCh)
+
+				glog.V(3).Info("waiting pv to detach for pod: ", pod.Name)
+				select {
+				case <-detachCh:
+					glog.V(3).Info("Detached pv for pod: ", pod.Name)
+					returnCh <- nil
+				case <-time.After(o.PvDetachTimeout - time.Now().Sub(startTime)):
+					glog.Warningf("Timout occured when detaching pv for pod: %v", pod.Name)
+					returnCh <- nil
+				}
+			case <-time.After(o.PvDetachTimeout):
+				returnCh <- fmt.Errorf("Timout when evicting pod: %v", pod)
+			}
+		}
+		if len(retryPods) == 0 {
+			break
+		}
+		glog.V(4).Info("Eviction for some pods will be retried after 5 seconds.")
+		pods = retryPods
+		time.Sleep(5 * time.Second)
+	}
+	return
+}
+
+func (o *DrainOptions) getPvs(pod corev1.Pod) []string {
```

Comment:
Why not return a deep structure including the volume ids (with a nested call to the drivers) and avoid repeated GET PVC call [below](https://github.com/gardener/machine-controller-manager/pull/275/files#diff-face99aebd3cd79577003f7df747c863R572)?